src/org/rascalmpl/uri/file/MavenRepositoryURIResolver.java

```diff
@@ -217,4 +225,39 @@ public ISourceLocation resolve(ISourceLocation input) throws IOException {
             }
         }    
     }
+
+    public static ISourceLocation make(String groupId, String artifactId, String version, String path) {
+        return URIUtil.correctLocation("mvn", groupId + "--" + artifactId + "--" + version, path);
+    }
+
+    /** 
+     * Shortens a location of a jar file that points into a local maven repository, and
+     * leaves all other locations as-is.
+     * */
+    public static ISourceLocation mavenize(ISourceLocation loc) {
+        try {
+            loc = URIResolverRegistry.getInstance().logicalToPhysical(loc);
+
+            if (!URIUtil.getExtension(loc).equals("jar")) {
+                return loc;
+            }
+
+            ISourceLocation repo = URIUtil.createFileLocation(inferMavenRepositoryLocation());
+            ISourceLocation relative = URIUtil.relativize(repo, loc);
+            boolean isFileInRepo = loc.getScheme().equals("file") && relative.getScheme().equals("relative");  
+
+            if (isFileInRepo) { 
+                String groupId    = URIUtil.getParentLocation(URIUtil.getParentLocation(URIUtil.getParentLocation(relative))).getPath().substring(1).replaceAll("/", ".");
+                String artifactId = URIUtil.getLocationName(URIUtil.getParentLocation(URIUtil.getParentLocation(relative)));
+                String version    = URIUtil.getLocationName(URIUtil.getParentLocation(relative));
```

Comment:
if we factor out the `getParentLocation` calls into one or more variable above, this would get a lot more readable.

```java
relative = URIUtil.getParentLocation(relative);
String version    = URIUtil.getLocationName(relative);
relative = URIUtil.getParentLocation(relative);
String artifactId = URIUtil.getLocationName(relative);
relative = URIUtil.getParentLocation(relative);
String groupId    = relative.getPath().substring(1).replaceAll("/", ".");
```

for example.

Reply:
yes good plan